File: src/app/types/jass-overrides/trigger.lua
local ____exports = {}
____exports.Trigger = {}
local Trigger = ____exports.Trigger
Trigger.name = "Trigger"
Trigger.__index = Trigger
Trigger.prototype = {}
Trigger.prototype.__index = Trigger.prototype
Trigger.prototype.constructor = Trigger
function Trigger.new(...)
    local self = setmetatable({}, Trigger.prototype)
    self:____constructor(...)
    return self
end
function Trigger.prototype.____constructor(self)
    self.nativeTrigger = CreateTrigger()
end
function Trigger.EvaluateCondition(self, func)
    local answer = false
    xpcall(
        function()
            answer = func(nil)
        end,
        function(err)
            self:printError(err)
        end
    )
    return answer
end
function Trigger.printError(self, err)
    print("FATAL ERROR")
    print(err)
end
function Trigger.prototype.AddAction(self, actionFunc)
    return TriggerAddAction(
        self.nativeTrigger,
        function() return ({
            xpcall(
                function() return actionFunc(nil) end,
                function(err) return ____exports.Trigger:printError(err) end
            )
        }) end
    )
end
function Trigger.prototype.RegisterTimerEvent(self, timeout, periodic)
    return TriggerRegisterTimerEvent(self.nativeTrigger, timeout, periodic)
end
function Trigger.prototype.RegisterTimerEventSingle(self, timeout)
    return TriggerRegisterTimerEventSingle(self.nativeTrigger, timeout)
end
function Trigger.prototype.RegisterTimerEventPeriodic(self, timeout)
    return TriggerRegisterTimerEventPeriodic(self.nativeTrigger, timeout)
end
function Trigger.prototype.RegisterPlayerStateEvent(self, whichPlayer, whichState, opcode, limitval)
    return TriggerRegisterPlayerStateEvent(self.nativeTrigger, whichPlayer, whichState, opcode, limitval)
end
function Trigger.prototype.RegisterDeathEvent(self, whichWidget)
    return TriggerRegisterDeathEvent(self.nativeTrigger, whichWidget)
end
function Trigger.prototype.RegisterDialogEventBJ(self, whichDialog)
    return TriggerRegisterDialogEventBJ(self.nativeTrigger, whichDialog)
end
function Trigger.prototype.RegisterEnterRectSimple(self, r)
    return TriggerRegisterEnterRectSimple(self.nativeTrigger, r)
end
function Trigger.prototype.AddCondition(self, func)
    return TriggerAddCondition(
        self.nativeTrigger,
        Condition(
            function() return ____exports.Trigger:EvaluateCondition(func) end
        )
    )
end
function Trigger.prototype.AddFilterFuncCondition(self, filter)
    return TriggerAddCondition(self.nativeTrigger, filter)
end
function Trigger.prototype.RegisterAnyUnitEventBJ(self, whichEvent)
    TriggerRegisterAnyUnitEventBJ(self.nativeTrigger, whichEvent)
end
function Trigger.prototype.RegisterPlayerChatEvent(self, whichPlayer, chatMessageToDetect, exactMatchOnly)
    return TriggerRegisterPlayerChatEvent(self.nativeTrigger, whichPlayer, chatMessageToDetect, exactMatchOnly)
end
function Trigger.prototype.RegisterPlayerUnitEventSimple(self, whichPlayer, whichEvent)
    return TriggerRegisterPlayerUnitEventSimple(self.nativeTrigger, whichPlayer, whichEvent)
end
return ____exports

File: src/app/types/jass-overrides/trigger.ts
export class Trigger {

    nativeTrigger: trigger;
    public isDestroyed: boolean = false;

    constructor() {
        this.nativeTrigger = CreateTrigger();
    }

    public static EvaluateCondition(func: () => boolean): boolean {
        let answer: boolean = false;

        xpcall(() => {
            answer = func();
        }, (err: any) => {
            this.printError(err);
        });

        return answer;
    }


    public static printError(err: any) {
        print("FATAL ERROR");
        print(err);
    }


    public AddAction(actionFunc: () => void): triggeraction {
        return TriggerAddAction(this.nativeTrigger, () => xpcall(() => actionFunc(), err => Trigger.printError(err)));
    }

    public RegisterTimerExpire(timer: timer): event {
        return TriggerRegisterTimerExpireEvent(this.nativeTrigger, timer);
    }

    public RegisterTimerEvent(timeout: number, periodic: boolean): event {
        return TriggerRegisterTimerEvent(this.nativeTrigger, timeout, periodic);
    }

    public RegisterTimerEventSingle(timeout: number): event {
        return TriggerRegisterTimerEventSingle(this.nativeTrigger, timeout);
    }

    public RegisterTimerEventPeriodic(timeout: number): event {
        return TriggerRegisterTimerEventPeriodic(this.nativeTrigger, timeout);
    }

    public RegisterPlayerStateEvent(whichPlayer: player, whichState: playerstate, opcode: limitop, limitval: number): event {
        return TriggerRegisterPlayerStateEvent(this.nativeTrigger, whichPlayer, whichState, opcode, limitval);
    }

    public RegisterUnitTakesDamage() {
        return TriggerRegisterAnyUnitEventBJ(this.nativeTrigger, EVENT_PLAYER_UNIT_DAMAGED);
    }

    public RegisterDeathEvent(whichWidget: widget): event {
        return TriggerRegisterDeathEvent(this.nativeTrigger, whichWidget);
    }

    public RegisterDialogEventBJ(whichDialog: dialog): event {
        return TriggerRegisterDialogEventBJ(this.nativeTrigger, whichDialog);
    }


    public RegisterEnterRectSimple(r: rect): event {
        return TriggerRegisterEnterRectSimple(this.nativeTrigger, r);
    }

    // public RegisterEnterRectangle(area: Rectangle): event {
    //     const rectangle: rect = Rect(area.minX, area.minY, area.maxX, area.maxY);
    //     const trigEvent: event = this.RegisterEnterRectSimple(rectangle);
    //     RemoveRect(rectangle);
    //     return trigEvent;
    // }

    public AddCondition(func: () => boolean): triggercondition {
        return TriggerAddCondition(this.nativeTrigger, Condition(() => Trigger.EvaluateCondition(func)));
    }

    public AddFilterFuncCondition(filter: filterfunc): triggercondition {
        return TriggerAddCondition(this.nativeTrigger, filter);
    }

    public RegisterAnyUnitEventBJ(whichEvent: playerunitevent): void {
        TriggerRegisterAnyUnitEventBJ(this.nativeTrigger, whichEvent);
    }

    public RegisterPlayerChatEvent(whichPlayer: player, chatMessageToDetect: string, exactMatchOnly: boolean): event {
        return TriggerRegisterPlayerChatEvent(this.nativeTrigger, whichPlayer, chatMessageToDetect, exactMatchOnly);
    }

    public RegisterPlayerUnitEventSimple(whichPlayer: player, whichEvent: playerunitevent): event {
        return TriggerRegisterPlayerUnitEventSimple(this.nativeTrigger, whichPlayer, whichEvent);
    }

    // public RegisterPlayerEventLeave(whichPlayer: AbstractPlayer): event {
    //     return TriggerRegisterPlayerEventLeave(this.nativeTrigger, whichPlayer.wcPlayer);
    // }


    public RegisterUnitIssuedOrder(whichUnit: unit, whichEvent: unitevent): event {
        return TriggerRegisterUnitEvent(this.nativeTrigger, whichUnit, whichEvent);
    }

    public destroy() {
        this.isDestroyed = true;
        DestroyTrigger(this.nativeTrigger);
    }
}